bundles/remote_services/discovery_zeroconf/src/discovery_zeroconf_announcer.c

```diff
@@ -0,0 +1,533 @@
+    discovery_zeroconf_announcer_t *announcer = calloc(1, sizeof(*announcer));
+    assert(announcer != NULL);
+    announcer->ctx = ctx;
+    announcer->logHelper = logHelper;
+    announcer->sharedRef = NULL;
+
+    announcer->eventFd = eventfd(0, 0);
+    if (announcer->eventFd < 0) {
+        status = CELIX_ERROR_MAKE(CELIX_FACILITY_CERRNO, errno);
+        celix_logHelper_fatal(logHelper, "Announcer: Failed to open event fd, %d.", errno);
+        goto eventfd_err;
+    }
+
+    status = celixThreadMutex_create(&announcer->mutex, NULL);
+    if (status != CELIX_SUCCESS) {
+        celix_logHelper_fatal(logHelper, "Announcer: Failed to create mutex, %d.", status);
+        goto mutex_err;
+    }
+
+    announcer->endpoints = celix_stringHashMap_create();
+    assert(announcer->endpoints != NULL);
+    announcer->revokedEndpoints = celix_arrayList_create();
+    assert(announcer->revokedEndpoints != NULL);
+
+    const char *fwUuid = celix_bundleContext_getProperty(ctx, OSGI_FRAMEWORK_FRAMEWORK_UUID, NULL);
+    if (fwUuid == NULL || strlen(fwUuid) >= sizeof(announcer->fwUuid)) {
+        status = CELIX_BUNDLE_EXCEPTION;
+        celix_logHelper_fatal(logHelper, "Announcer: Failed to get framework uuid.");
+        goto fw_uuid_err;
+    }
+    strcpy(announcer->fwUuid, fwUuid);
+
+    announcer->epListener.handle = announcer;
+    announcer->epListener.endpointAdded = discoveryZeroconfAnnouncer_endpointAdded;
+    announcer->epListener.endpointRemoved = discoveryZeroconfAnnouncer_endpointRemoved;
+    char scope[256] = {0};
+    (void)snprintf(scope, sizeof(scope), "(&(%s=*)(%s=%s))", OSGI_FRAMEWORK_OBJECTCLASS, OSGI_RSA_ENDPOINT_FRAMEWORK_UUID, fwUuid);
+
+    celix_properties_t *props = celix_properties_create();
+    assert(props != NULL);
+    celix_properties_set(props, "DISCOVERY", "true");
+    celix_properties_set(props, (char *) OSGI_ENDPOINT_LISTENER_SCOPE, scope);
+    celix_service_registration_options_t opt = CELIX_EMPTY_SERVICE_REGISTRATION_OPTIONS;
+    opt.serviceName = OSGI_ENDPOINT_LISTENER_SERVICE;
+    opt.properties = props;
+    opt.svc = &announcer->epListener;
+    announcer->epListenerSvcId = celix_bundleContext_registerServiceWithOptionsAsync(ctx, &opt);
+    if (announcer->epListenerSvcId < 0) {
+        status = CELIX_BUNDLE_EXCEPTION;
+        celix_logHelper_fatal(logHelper, "Announcer: Failed to register endpoint listener.");
+        goto ep_listener_err;
+    }
+
+    announcer->running = true;
+    status = celixThread_create(&announcer->refreshEPThread, NULL, discoveryZeroconfAnnouncer_refreshEndpointThread, announcer);
+    if (status != CELIX_SUCCESS) {
+        celix_logHelper_fatal(logHelper, "Announcer: Failed to create refreshing endpoint thread, %d.", status);
+        goto announce_ep_thread_err;
+    }
+    celixThread_setName(&announcer->refreshEPThread, "DiscAnnouncer");
+
+    *announcerOut = announcer;
+    return CELIX_SUCCESS;
+announce_ep_thread_err:
```

Comment:
I do think the usage of goto is better for error handler than lots of if/else and I also think it would be nice the document how we prefer to do error handling for future code.
But this is for another pull request.

Reply:
In my daily work, peng recommends me to use the usage. I think it's a better guarantee to release resources in reverse order.

Reply:
> I do think the usage of goto is better for error handler than lots of if/else and I also think it would be nice the document how we prefer to do error handling for future code.

When writing C in OOP-style, this might be the best we can do to mimic ctor/dtor. 
`__attribute__((cleanup))` is interesting, but I have not used it in large scale.

Reply:
My first though is that this is cleaner than gotos, but the are also some downsides:
 - The cleanup callback is expected to take a `void*` arg, so you loose type safety.
 - The cleanup attribute is a GNU extension (I think also supported by clang) and not part of C99 or C11.
 - Maybe requires quite some extra (static) functions, which can made the code more but also less readable. 
 - IMO attribute usage if verbose, which also impact the readability.

Nonetheless interesting and maybe something to try out.